README.rst
===============================
Asyncio HTTP Server
===============================

.. image:: https://img.shields.io/travis/thomwiggers/httpserver.svg
        :target: https://travis-ci.org/thomwiggers/httpserver

.. image:: https://img.shields.io/pypi/v/httpserver.svg
        :target: https://pypi.python.org/pypi/httpserver

Python 3.4 asyncio implementation of an HTTP server

* Free software: BSD license
* Documentation: http://pythonhosted.org/httpserver/

Features
--------

* Supports HTTP/1.1
* Supports Keep-Alive
* Serve a specific hostname only
* Bind to an ip address
* Uses asyncio
* Installs an 'httpserver' command


HISTORY.rst
.. :changelog:

History
-------

1.1.0 (2015-04-02)
---------------------

* No longer rely on sockets to timeout themselves
* Add default timeout of 15 seconds to httpserver
* Negotiate with the Keep-Alive header
* Add ``--version`` command
* Change documentation theme to ``alabaster``

1.0.1 (2015-03-21)
---------------------
* Fix small issue in README on pypi

1.0.0 (2015-03-21)
---------------------

* First release


tests/test_selenium.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_selenium
----------------------------------

Tests the http server with a real browser.
"""
import os
import unittest
import multiprocessing
import time

from httpserver import _start_server

from selenium import webdriver

_host = 'localhost'
_ip = '127.0.0.1'
_port = 1234
_dir = os.path.join(os.path.dirname(__file__), 'fixtures/selenium')


class TestSelenium(unittest.TestCase):

    def _sleep(self):
        if not isinstance(self.driver, webdriver.PhantomJS):
            time.sleep(3)

    def setUp(self):
        """Set up selenium"""
        super(TestSelenium, self).setUp()
        driver = os.environ.get("SELENIUM_WEBDRIVER", "Firefox")
        if driver == "phantomjs":
            self.driver = webdriver.PhantomJS()
        elif driver == "chrome":
            self.driver = webdriver.Chrome()
        else:
            self.driver = webdriver.Firefox()
        self.driver.implicitly_wait(3)

        self.process = multiprocessing.Process(
            target=_start_server,
            args=(_ip, _port, _host, _dir))
        self.process.start()

    def test_get_index(self):
        hostname = 'http://{}:{}'.format(_host, _port)
        self.driver.get(hostname)
        self.driver.find_elements_by_tag_name('title')
        self._sleep()
        assert '<title>Index</title>' in self.driver.page_source

    def test_get_404(self):
        path = 'http://{}:{}/notfound'.format(_host, _port)
        self.driver.get(path)
        self._sleep()
        assert 'Not Found' in self.driver.page_source

    def tearDown(self):
        super(TestSelenium, self).tearDown()
        self.process.terminate()
        self.driver.quit()

if __name__ == '__main__':
    unittest.main()


tests/test_httpserver.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_httpserver
----------------------------------

Tests for `httpserver` module.
"""
import os
import unittest
import hashlib
from unittest import mock
from freezegun import freeze_time

from httpserver.httpserver import HttpProtocol


class TestHttpserver(unittest.TestCase):

    @mock.patch('asyncio.get_event_loop')
    def setUp(self, patch):
        """Set up an HttpProtocol instance with a fake transport"""
        self.fixtures_location = os.path.join(
            os.path.dirname(__file__), 'fixtures')
        self.httpprotocol = HttpProtocol('localhost', self.fixtures_location)
        self.transport = mock.MagicMock(spec=['write', 'close'])
        self.transport.write = mock.Mock()
        self.transport.close = mock.Mock()
        self.httpprotocol.connection_made(self.transport)

    def _read_fixture(self, filename):
        """Convenient function to read one of our fixtures"""
        filename = os.path.join(self.fixtures_location, filename)
        with open(filename, 'rb') as f:
            content = f.read()
        return content

    def _sent(self):
        """Gets the data that was written to the transport"""
        values = [tuple(c)[0][0]
                  for c in self.transport.write.call_args_list]
        self.transport.write.reset_mock()
        return b''.join(values)

    def test_invalid_method(self):
        """Test if the server rejects methods it does not recognise"""
        data = self._read_fixture('invalid_method.crlf')
        self.httpprotocol.data_received(data)

        response = self._sent()
        # Do we have the error?
        assert response.startswith(b'HTTP/1.1 501 Not Implemented\r\n')
        head, body = response.split(b'\r\n\r\n', 1)
        assert b'Content-Length' in head

        # Has the connection been closed?
        assert self.transport.close.called

    def test_invalid_http_version(self):
        """Test if the server rejects invalid HTTP headers"""
        data = self._read_fixture('invalid_version.crlf')
        self.httpprotocol.data_received(data)
        response = self._sent()

        head, body = response.split(b'\r\n\r\n', 1)
        # Do we have the error
        assert head.startswith(b'HTTP/1.1 505 HTTP Version Not Supported')
        assert b'Content-Length' in head  # We SHOULD have an entity (10.5.6)
        assert b'HTTP/1.1' in body  # We SHOULD reply what is supported

    def test_get_index_root(self):
        """Test GET /"""
        data = self._read_fixture('get_index_root.crlf')
        index = self._read_fixture('index.html')
        self.httpprotocol.data_received(data)

        response = self._sent()
        assert response.startswith(b'HTTP/1.1 200 OK\r\n')
        head, body = response.split(b'\r\n\r\n', 1)
        assert 'Content-Length: {}'.format(len(body)).encode('utf-8') in head
        assert body == index

    def test_get_index_named(self):
        """Test GET /index.html"""
        data = self._read_fixture('get_index_named.crlf')
        index = self._read_fixture('index.html')
        self.httpprotocol.data_received(data)
        response = self._sent()
        assert response.startswith(b'HTTP/1.1 200 OK\r\n')
        head, body = response.split(b'\r\n\r\n', 1)
        assert 'Content-Length: {}'.format(len(body)).encode('utf-8') in head
        assert body == index

        # Has the connection been closed (no Keep-Alive)
        assert self.transport.close.called

    def test_get_directory_without_index(self):
        """Test GET /emptydir"""
        data = self._read_fixture(
            'no_index_dir/get_directory_without_index.crlf')
        self.httpprotocol.data_received(data)
        response = self._sent()

        head, body = response.split(b'\r\n\r\n', 1)
        # Do we have the 404 error
        assert head.startswith(b'HTTP/1.1 404 Not Found\r\n')
        assert self.transport.close.called

    def test_get_absoluteURI(self):
        """HTTP 1.1 servers MUST accept absoluteURI form Request-URIs

        RFC 2616 section 5.1.2
        """
        data = self._read_fixture('absolute_uri_request.crlf')
        self.httpprotocol.data_received(data)
        head, body = self._sent().split(b'\r\n\r\n', 1)

        assert head.startswith(b'HTTP/1.1 200 OK\r\n')

    def test_get_absoluteURI_404(self):
        """We don't want to serve our index.html on GET nu.nl"""
        data = self._read_fixture('absolute_uri_wrong_host.crlf')
        self.httpprotocol.data_received(data)
        head, body = self._sent().split(b'\r\n\r\n', 1)

        assert head.startswith(b'HTTP/1.1 404 Not Found\r\n')

    def test_get_persistent(self):
        """Try and get index.html over the same connection"""
        data = self._read_fixture('get_index_persistent.crlf')
        index = self._read_fixture('index.html')

        for i in range(2):
            self.httpprotocol.data_received(data)
            head, body = self._sent().split(b'\r\n\r\n', 1)
            assert head.startswith(b'HTTP/1.1 200 OK')
            assert b'Keep-Alive' in head
            assert body == index
            assert not self.transport.close.called

        data = self._read_fixture('get_index_named.crlf')
        self.httpprotocol.data_received(data)
        head, body = self._sent().split(b'\r\n\r\n', 1)
        assert head.startswith(b'HTTP/1.1 200 OK')
        assert body == index
        assert self.transport.close.called

    @freeze_time("2015-03-14 09:26:53", tz_offset=0)
    def test_get_has_date_headers(self):
        """Test for date headers"""
        requests = [
            self._read_fixture('get_index_named.crlf'),  # 200
            self._read_fixture(
                'no_index_dir/get_directory_without_index.crlf'),  # 404
            self._read_fixture('invalid_version.crlf'),  # 505
        ]

        for request in requests:
            self.httpprotocol.data_received(request)
            response = self._sent()

            head, body = response.split(b'\r\n\r\n', 1)

            assert b'Date: ' in head
            assert b'Date: Sat, 14 Mar 2015 09:26:53 +0000' in head

    def test_get_with_etag(self):
        """Test GET /index.html"""
        request = self._read_fixture('get_index_named.crlf')
        index = self._read_fixture('index.html')
        self.httpprotocol.data_received(request)
        response = self._sent()
        head, body = response.split(b'\r\n\r\n', 1)

        sha1 = hashlib.sha1()
        sha1.update(body)
        etag = sha1.hexdigest()
        sha1 = hashlib.sha1()
        sha1.update(index)
        etag_index = sha1.hexdigest()

        assert 'Etag: "{}"'.format(etag).encode('utf-8') in head
        assert etag == etag_index

        request = self._read_fixture('get_index_if_none_match.crlf')
        assert etag.encode('utf-8') in request

        self.httpprotocol.data_received(request)
        response = self._sent()
        head, body = response.split(b'\r\n\r\n', 1)

        assert head.startswith(b'HTTP/1.1 304 Not Modified\r\n')
        assert b'Date: ' in head
        assert 'Etag: "{}"'.format(etag).encode('utf-8') in head

    def test_send_keepalive(self):
        """Send a lower keepalive"""
        request = self._read_fixture('get_keepalive_2.crlf')
        self.httpprotocol.data_received(request)
        response = self._sent()
        head, body = response.split(b'\r\n\r\n', 1)
        assert b'timeout=2' in head
        assert self.httpprotocol._timeout == 2

    def test_send_keepalive_longer(self):
        """Server sends a lower timeout if we request a too long one"""
        request = self._read_fixture('get_keepalive_2000.crlf')
        self.httpprotocol.data_received(request)
        response = self._sent()
        head, body = response.split(b'\r\n\r\n', 1)
        assert b'timeout=15' in head
        assert self.httpprotocol._timeout == 15

    def test_timeout(self):
        """Test timing out"""
        assert not self.transport.close.called
        # Test if the handle_timeout event closes the socket
        # actual waiting is not possible because we mock the transport
        # and don't start an event loop.
        self.httpprotocol._handle_timeout()
        assert self.transport.close.called

    def test_timeout_request(self):
        data = self._read_fixture('get_index_persistent.crlf')
        self.httpprotocol.data_received(data)
        assert not self.transport.close.called
        # Test if the handle_timeout event closes the socket
        self.httpprotocol._handle_timeout()
        assert self.transport.close.called


if __name__ == '__main__':
    unittest.main()


httpserver/__init__.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals, absolute_import, print_function

import logging

__author__ = 'Thom Wiggers, Luuk Scholten'
__email__ = 'thom@thomwiggers.nl, info@luukscholten.com'
__version__ = '1.1.0'


def _start_server(bindaddr, port, hostname, folder):
    """Starts an asyncio server"""
    import asyncio
    from .httpserver import HttpProtocol
    loop = asyncio.get_event_loop()
    coroutine = loop.create_server(lambda: HttpProtocol(hostname, folder),
                                   bindaddr,
                                   port)
    server = loop.run_until_complete(coroutine)

    print('Starting server on {}'.format(server.sockets[0].getsockname()))
    try:
        loop.run_forever()
    except KeyboardInterrupt:
        pass


def run(argv=None):  # pragma: no cover
    """Run the HTTP server

    Usage:
        httpserver [options] [<folder>]

    Options::

        -h,--host=<hostname>        What host name to serve (default localhost)
        -a,--bindaddress=<address>  Address to bind to (default 127.0.0.1)
        -p,--port=<port>            Port to listen on (default 8080)
        -v,--verbose                Increase verbosity to INFO messages
        -d,--debug                  Increase verbosity to DEBUG messages
        --help                      Print this help message
        --version                   Print the version

    To serve /path/to/www on all (ipv4) addresses for host myserver
    on port 80::
        httpserver -a 0.0.0.0 -p 80 -h myserver /path/to/www
    """
    import sys
    import os
    import docopt
    import textwrap

    # Check for the version
    if not sys.version_info >= (3, 4):
        print('This python version is not supported. Please use python 3.4')
        exit(1)

    argv = argv or sys.argv[1:]

    # remove some RST formatting
    docblock = run.__doc__.replace('::', ':')
    args = docopt.docopt(textwrap.dedent(docblock), argv)

    if args['--version']:
        print("httpserver version {} by {}".format(
            __version__,
            __author__))
        exit(0)

    # Set up logging
    level = logging.WARNING
    if args['--verbose']:
        level = logging.INFO
    if args['--debug']:
        level = logging.DEBUG
    logging.basicConfig(level=level)

    logger = logging.getLogger('run method')
    logger.debug('CLI args: %s' % args)

    bindaddr = args['--bindaddress'] or '127.0.0.1'
    port = args['--port'] or '8080'
    folder = args['<folder>'] or os.getcwd()
    hostname = args['--host'] or 'localhost'
    _start_server(bindaddr, port, hostname, folder)


httpserver/httpserver.py
# -*- coding: utf-8 -*-
import asyncio
import hashlib
import logging
import mimetypes
import os
import re
from datetime import datetime
from http.client import responses
from urllib.parse import unquote


logger = logging.getLogger(__name__)


def _get_response(**kwargs):
    """Get a template response

    Use kwargs to add things to the dictionary
    """
    if 'code' not in kwargs:
        kwargs['code'] = 200
    if 'headers' not in kwargs:
        kwargs['headers'] = dict()
    if 'version' not in kwargs:
        kwargs['version'] = 'HTTP/1.1'

    return dict(**kwargs)


class HttpProtocol(asyncio.Protocol):
    """HTTP/1.1 Protocol implementation

    Per connection made, one of these gets instantiated
    """

    def __init__(self, host, folder, event_loop=None, timeout=15):
        """Initialise a new instance.

        Arguments:
            host: the host to serve
            folder: the folder to serve files from
        """
        self.host = host
        self.folder = folder
        self.logger = logger.getChild('HttpProtocol {}'.format(id(self)))
        self.logger.debug('Instantiated HttpProtocol')
        self._loop = event_loop or asyncio.get_event_loop()
        self._timeout = timeout
        self._timeout_handle = None

    def _write_transport(self, string):
        """Convenience function to write to the transport"""
        if isinstance(string, str):  # we need to convert to bytes
            self.transport.write(string.encode('utf-8'))
        else:
            self.transport.write(string)

    def _write_response(self, response):
        """Write the response back to the client

        Arguments:
        response -- the dictionary containing the response.
        """
        status = '{} {} {}\r\n'.format(response['version'],
                                       response['code'],
                                       responses[response['code']])
        self.logger.debug("Responding status: '%s'", status.strip())
        self._write_transport(status)

        if 'body' in response and 'Content-Length' not in response['headers']:
            response['headers']['Content-Length'] = len(response['body'])

        response['headers']['Date'] = datetime.utcnow().strftime(
            "%a, %d %b %Y %H:%M:%S +0000")

        for (header, content) in response['headers'].items():
            self.logger.debug("Sending header: '%s: %s'", header, content)
            self._write_transport('{}: {}\r\n'.format(header, content))

        self._write_transport('\r\n')
        if 'body' in response:
            self._write_transport(response['body'])

    def connection_made(self, transport):
        """Called when the connection is made"""
        self.logger.info('Connection made at object %s', id(self))
        self.transport = transport
        self.keepalive = True

        if self._timeout:
            self.logger.debug('Registering timeout event')
            self._timout_handle = self._loop.call_later(
                self._timeout, self._handle_timeout)

    def connection_lost(self, exception):
        """Called when the connection is lost or closed.

        The argument is either an exception object or None. The latter means
        a regular EOF is received, or the connection was aborted or closed by
        this side of the connection.
        """
        if exception:
            self.logger.exception('Connection lost!')
        else:
            self.logger.info('Connection lost')

    def data_received(self, data):
        """Process received data from the socket

        Called when we receive data
        """
        self.logger.debug('Received data: %s', repr(data))

        try:
            request = self._parse_headers(data)
            self._handle_request(request)
        except InvalidRequestError as e:
            self._write_response(e.get_http_response())

        if not self.keepalive:
            if self._timeout_handle:
                self._timeout_handle.cancel()
            self.transport.close()

        if self._timeout and self._timeout_handle:
            self.logger.debug('Delaying timeout event')
            self._timeout_handle.cancel()
            self._timout_handle = self._loop.call_later(
                self._timeout, self._handle_timeout)

    def _parse_headers(self, data):
        self.logger.debug('Parsing headers')

        request_strings = list(map(lambda x: x.decode('utf-8'),
                                   data.split(b'\r\n')))

        request = dict()

        # Parse request method and HTTP version
        method_line = request_strings[0].split()

        # The first line has either 3 or 2 arguments
        if not (2 <= len(method_line) <= 3):
            self.logger.info('Got an invalid http header')
            self.keepalive = False  # We don't trust you
            raise InvalidRequestError(400, 'Bad request')
        # HTTP 0.9 isn't supported.
        if len(method_line) == 2:
            self.logger.info('Got a HTTP/0.9 request')
            self.keepalive = False  # HTTP/0.9 won't support persistence
            raise InvalidRequestError(505, "This server only supports HTTP/1.0"
                                           "and HTTP/1.1")
        else:
            request['version'] = method_line[2]

        # method
        request['method'] = method_line[0]
        # request URI
        request['target'] = method_line[1]

        # Parse the headers
        for line in request_strings[1:]:
            if line == '':  # an empty line signals the end of the headers
                break
            self.logger.debug("header: '%s'", line)
            header, body = line.split(': ', 1)
            request[header] = body

        self.logger.debug('request object: %s', request)
        return request

    def _get_request_uri(self, request):
        """Parse the request URI into something useful

        Server MUST accept full URIs (5.1.2)"""
        request_uri = request['target']
        if request_uri.startswith('/'):  # eg. GET /index.html
            return (request.get('Host', 'localhost').split(':')[0],
                    request_uri[1:])
        elif '://' in request_uri:  # eg. GET http://rded.nl
            locator = request_uri.split('://', 1)[1]
            host, path = locator.split('/', 1)
            return (host.split(':')[0], path)

    def _handle_request(self, request):
        """Process the headers and get the file"""

        # Check if this is a persistent connection.
        if request['version'] == 'HTTP/1.1':
            self.keepalive = not request.get('Connection') == 'close'
        elif request['version'] == 'HTTP/1.0':
            self.keepalive = request.get('Connection') == 'Keep-Alive'

        # Check if we're getting a sane request
        if request['method'] not in ('GET'):
            raise InvalidRequestError(501, 'Method not implemented')
        if request['version'] not in ('HTTP/1.0', 'HTTP/1.1'):
            raise InvalidRequestError(
                505, 'Version not supported. Supported versions are: {}, {}'
                .format('HTTP/1.0', 'HTTP/1.1'))

        host, location = self._get_request_uri(request)

        # We must ignore the Host header if a host is specified in GET
        if host is None:
            host = request.get('Host')

        # Check if this request is intended for this webserver
        if host is not None and not host == self.host:
            self.logger.info('Got a request for unknown host %s', host)
            raise InvalidRequestError(404, "We don't serve this host")

        filename = os.path.join(self.folder, unquote(location))
        self.logger.debug('trying to serve %s', filename)

        if os.path.isdir(filename):
            filename = os.path.join(filename, 'index.html')

        if not os.path.isfile(filename):
            raise InvalidRequestError(404, 'Not Found')

        # Start response with version
        response = _get_response(version=request['version'])

        # timeout negotiation
        match = re.match(r'timeout=(\d+)', request.get('Keep-Alive', ''))
        if match is not None:
            requested_timeout = int(match.group(1))
            if requested_timeout < self._timeout:
                self._timeout = requested_timeout

        # tell the client our timeout
        if self.keepalive:
            response['headers'][
                'Keep-Alive'] = 'timeout={}'.format(self._timeout)

        # Set Content-Type
        response['headers']['Content-Type'] = mimetypes.guess_type(
            filename)[0] or 'text/plain'

        # Generate E-tag
        sha1 = hashlib.sha1()
        with open(filename, 'rb') as fp:
            response['body'] = fp.read()
            sha1.update(response['body'])
        etag = sha1.hexdigest()

        # Create 304 response if if-none-match matches etag
        if request.get('If-None-Match') == '"{}"'.format(etag):
            # 304 responses shouldn't contain many headers we might already
            # have added.
            response = _get_response(code=304)

        response['headers']['Etag'] = '"{}"'.format(etag)

        self._write_response(response)

    def _handle_timeout(self):
        """Handle a timeout"""
        self.logger.info('Request timed out')
        self.transport.close()


class InvalidRequestError(Exception):
    """Raised for invalid requests. Contains the error code.

    This exception can be transformed to a http response.
    """

    def __init__(self, code, *args, **kwargs):
        """Configures a new InvalidRequestError

        Arguments:
            code -- the HTTP error code
        """
        super(InvalidRequestError, self).__init__(*args, **kwargs)
        self.code = code

    def get_http_response(self):
        """Get this exception as an HTTP response suitable for output"""
        return _get_response(
            code=self.code,
            body=str(self),
            headers={
                'Content-Type': 'text/plain'
            }
        )
